pkg/gofr/datasource/opentsdb/opentsdb.go

```diff
@@ -0,0 +1,353 @@
+	span := c.addTrace(c.ctx, "HealthCheck")
+
+	status := StatusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "HealthCheck", &status, &message, span)
+
+	h := Health{
+		Details: make(map[string]interface{}),
+	}
+
+	conn, err := dialTimeout("tcp", c.opentsdbCfg.OpentsdbHost, DefaultDialTime)
+	if err != nil {
+		h.Status = "DOWN"
+		message = fmt.Sprintf("OpenTSDB is unreachable: %v", err)
+
+		return nil, errors.New(message)
+	}
+
+	if conn != nil {
+		defer conn.Close()
+	}
+
+	h.Details["host"] = c.tsdbEndpoint
+
+	ver, err := c.version()
+	if err != nil {
+		message = err.Error()
+		return nil, err
+	}
+
+	h.Details["version"] = ver.VersionInfo["version"]
+
+	status = StatusSuccess
+	h.Status = "UP"
+	message = "connection to OpenTSDB is alive"
+
+	return &h, nil
+}
+
+// sendRequest dispatches an HTTP request to the OpenTSDB server, using the provided
+// method, URL, and body content. It returns the parsed response or an error, if any.
+func (c *Client) sendRequest(method, url, reqBodyCnt string, parsedResp Response) error {
+	span := c.addTrace(c.ctx, "sendRequest")
+
+	status := StatusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "sendRequest", &status, &message, span)
+
+	// Create the HTTP request, attaching the context if available.
+	req, err := http.NewRequest(method, url, strings.NewReader(reqBodyCnt))
+	if c.ctx != nil {
+		req = req.WithContext(c.ctx)
+	}
+
+	if err != nil {
+		errRequestCreation := fmt.Errorf("failed to create request for %s %s: %v", method, url, err)
+
+		message = fmt.Sprint(errRequestCreation)
+
+		return errRequestCreation
+	}
+
+	// Set the request headers.
+	req.Header.Set("Content-Type", "application/json; charset=UTF-8")
+
+	// Send the request and handle the response.
+	resp, err := c.client.Do(req)
+	if err != nil {
+		errSendingRequest := fmt.Errorf("failed to send request for %s %s: %v", method, url, err)
+
+		message = fmt.Sprint(errSendingRequest)
+
+		return errSendingRequest
+	}
+
+	defer resp.Body.Close()
+
+	// Read and parse the response.
+	jsonBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		errReading := fmt.Errorf("failed to read response body for %s %s: %v", method, url, err)
+
+		message = fmt.Sprint(errReading)
+
+		return errReading
+	}
+
+	parsedResp.SetStatus(resp.StatusCode)
+
+	parser := parsedResp.GetCustomParser()
+	if parser == nil {
+		// Use the default JSON unmarshaller if no custom parser is provided.
+		if err := json.Unmarshal(jsonBytes, parsedResp); err != nil {
+			errUnmarshaling := fmt.Errorf("failed to unmarshal response body for %s %s: %v", method, url, err)
+
+			message = fmt.Sprint(errUnmarshaling)
+
+			return errUnmarshaling
+		}
+	} else {
+		// Use the custom parser if available.
+		if err := parser(jsonBytes); err != nil {
+			message = fmt.Sprintf("failed to parse response body through custom parser %s %s: %v", method, url, err)
```

Comment:
```suggestion
			message = fmt.Sprintf("failed to parse response body through custom parser %s %s: %w", method, url, err)
```

Reply:
"%w" is used with fmt.Errorf() calls....